daemon/snapper/timestamp_formatting/timestamp_formatting.go

```diff
@@ -0,0 +1,92 @@
+package timestamp_formatting
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+type Formatter struct {
+	format   func(time.Time) string
+	location *time.Location
+}
+
+func New(formatString string, locationString string) (*Formatter, error) {
+	location, err := time.LoadLocation(locationString) // no shadow
+	if err != nil {
+		return nil, errors.Wrapf(err, "load location from string %q", locationString)
+	}
+	makeFormatFunc := func(formatString string) (func(time.Time) string, error) {
+		if strings.Contains(formatString, "+") {
+			return nil, fmt.Errorf("character '+' is not allowed in ZFS snapshot names")
+		}
+		return func(t time.Time) string {
+			res := t.Format(formatString)
+			// if the formatString contains a time zone specifier
+			// and the location would result in a positive offset to UTC
+			// then the result of t.Format would contain a '+' sign.
+			if isLocationPositiveOffsetToUTC(location) {
+				// the only source of `+` can be the positive time zone offset because we disallowed `+` as a character in the format string
+				res = strings.Replace(res, "+", "_", 1)
+			}
+			if strings.Contains(res, "+") {
+				panic(fmt.Sprintf("format produced a string containing illegal character '+' that wasn't the expected case of positive time zone offset: format=%q location=%q unix=%q result=%q", formatString, location, t.Unix(), res))
+			}
+			return res
+		}, nil
+	}
+	var formatFunc func(time.Time) string
+	mustUseUtcError := func() error {
+		return fmt.Errorf("format string requires UTC location")
+	}
+	switch strings.ToLower(formatString) {
+	case "dense":
```

Comment:
In this case, I would actually print a warning maybe (or maybe not) that location is not UTC and force it to be UTC in code, and not give an error, because predefined format masks require implicit use of UTC.

Reply:
> and not give an error,

The error will force the user to recognize that their configuration is incorrect.
The default `timestamp_location` will be UTC, so,

```
timestamp_format: dense
```

will continue to work without error.

Only if they explicitly specify something wrong, like,

```yaml
timestamp_format: dense
timestamp_location: Europe/Berlin
```

will there be an error.